alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+        if num_prototypes > len(self.Y):
+            num_prototypes = len(self.Y)
+            logger.warning('The number of prototypes requested is larger than the number of elements from '
+                           f'the prototypes selection set. Automatically setting `num_prototypes={num_prototypes}`.')
+
+        # dictionary of prototypes indices for each class
+        protos = {l: [] for l in range(self.max_label + 1)}  # type: Dict[int, List[int]]  # noqa: E741
+        # set of available prototypes indices. Note that initially we start with the entire set of Y,
+        # but as the algorithm progresses, we remove the indices of the prototypes that we already selected.
+        available_indices = set(range(len(self.Y)))
+        # matrix of size `[NY, NX]`, where `NY = len(Y)` and `NX = len(X)`
+        # represents a mask which indicates for each element `y` in `Y` what are the elements of `X` that are in an
+        # epsilon ball centered in `y`.
+        B = (self.kmatrix_yx <= self.eps).astype(np.int32)
+        # matrix of size `[L, NX]`, where `L` is the number of labels
+        # each row `l` indicates the elements from `X` that are covered by prototypes belonging to class `l`
+        B_P = np.zeros((self.max_label + 1, len(self.X)), dtype=np.int32)
+        # matrix of size `[L, NX]`. Each row `l` indicates which elements form `X` are labeled as `l`
+        Xl = np.concatenate([(self.X_labels == l).reshape(1, -1)
+                             for l in range(self.max_label + 1)], axis=0).astype(np.int32)  # noqa: E741
+
+        # vectorized implementation of the prototypes scores.
+        # See paper (pag 8): https://arxiv.org/pdf/1202.5933.pdf for more details
+        B_diff = B[:, np.newaxis, :] - B_P[np.newaxis, :, :]  # [NY, 1, NX] - [1, L, NX] -> [NY, L, NX]
+        # [NY, L, NX] + [1, L, NX] -> [NY, L, NX]
+        delta_xi_all = B_diff + Xl[np.newaxis, ...] >= 2
+        # [NY, L]. For every row `y` and every column `l`, we compute how many new instances belonging to class
+        # `l` will be covered if we add the prototype `y`.
+        delta_xi_summed = np.sum(delta_xi_all, axis=-1)
+        # [NY, 1, NX] +  [1, L, NX] -> [NY, L, NX]
+        delta_nu_all = B[:, np.newaxis, :] + (1 - Xl[np.newaxis, ...]) >= 2
+        # [NY, L]. For every row `y` and every column `l`, we compute how many new instances belonging to all the
+        # other classes different then `l` will be covered if we add the prototype `y`.
+        delta_nu_summed = np.sum(delta_nu_all, axis=-1)
+        # compute the tradeoff score - each prototype tries to cover as many new elements as possible
+        # belonging to the same class, while trying to avoid covering elements belonging to another class
+        scores_all = delta_xi_summed - delta_nu_summed - self.lbd
+
+        # add progressing bar if `verbose=True`
+        generator = range(num_prototypes)
+        if self.verbose:
+            generator = tqdm(generator)
```

Comment:
Could it be cleaner to use the `disable` parameter to `tqdm` like Alex did in #628? I.e. `generator = tqdm(generator, disable=not verbose)` ?